File: Python/AStar_Vincent.py
from queue import PriorityQueue

class Edge:
    def __init__(self, weight, node1_index, node2_index):
        self.weight = weight
        self.node1_index = node1_index
        self.node2_index = node2_index

class Node:
    def __init__(self, name, index, hCost):
        self.name = name
        self.index = index
        self.gCost = 1000000000
        self.hCost = hCost
        self.parent_index = -1

class Graph:
    def __init__(self):
        self.nodes = []
        self.edges = []

    def addNode(self, name, hCost):
        self.nodes.append(Node(name, len(self.nodes), hCost))
        return len(self.nodes) - 1

    def addEdge(self, weight, node1_index, node2_index):
        self.edges.append(Edge(weight, node1_index, node2_index))

    def getEdges(self, node_index):
        edges = []
        for edge in self.edges:
            if edge.node1_index == node_index:
                edges.append(edge)
            elif edge.node2_index == node_index:
                edges.append(edge)
        return edges

    def getNeighbours(self, node_index):
        neighbours = []
        for edge in self.edges:
            if edge.node1_index == node_index:
                neighbours.append(self.nodes[edge.node2_index])
            elif edge.node2_index == node_index:
                neighbours.append(self.nodes[edge.node1_index])
        return neighbours

    def getEdge(self, node1_index, node2_index):
        for edge in self.edges:
            if edge.node1_index == node1_index and edge.node2_index == node2_index:
                return edge
            elif edge.node1_index == node2_index and edge.node2_index == node1_index:
                return edge
        return None

    def getEdgeWeight(self, node1_index, node2_index):
        edge = self.getEdge(node1_index, node2_index)
        if edge is None:
            return None
        return edge.weight

    def getNode(self, node_index):
        for node in self.nodes:
            if node.index == node_index:
                return node
        return None

    def print(self):
        for node in self.nodes:
            print(node.name, ":")
            for neighbour in self.getNeighbours(node.index):
                print("    ", neighbour.name)

class Path:
    def __init__(self, weight):
        self.weight = weight
        self.nodes = []

    def addNode(self, node_index):
        self.nodes.append(node_index)

    def print(self, graph):
        print("Path:")
        for node in self.nodes:
            print("    ", graph.getNode(node).name)

def aStar(graph, start_index, end_index):
    queue = PriorityQueue()
    visited = []
    startNode = graph.getNode(start_index)
    startNode.gCost = 0
    queue.put((0, startNode))

    while not queue.empty():
        currentNode = queue.get()[1]
        if currentNode.index == end_index:
            break
        visited.append(currentNode.index)
        for neighbour in graph.getNeighbours(currentNode.index):
            if neighbour.index in visited:
                continue
            edgeWeight = graph.getEdgeWeight(currentNode.index, neighbour.index)
            if edgeWeight is None:
                continue
            newGCost = currentNode.gCost + edgeWeight
            if newGCost < neighbour.gCost:
                neighbour.gCost = newGCost
                neighbour.parent_index = currentNode.index
                queue.put((neighbour.gCost + neighbour.hCost, neighbour))
    
    path = Path(graph.getNode(end_index).gCost)
    path.addNode(end_index)
    currentNode = graph.getNode(end_index)
    while currentNode.parent_index != -1:
        path.addNode(currentNode.parent_index)
        currentNode = graph.getNode(currentNode.parent_index)
    path.nodes.reverse()
    return path



graph = Graph()
a = graph.addNode("A", 10)
b = graph.addNode("B", 3)
c = graph.addNode("C", 5)
d = graph.addNode("D", 6)
e = graph.addNode("E", 8)
f = graph.addNode("F", 4)
g = graph.addNode("G", 2)
h = graph.addNode("H", 0)

graph.addEdge(2, a, b)
graph.addEdge(10, a, c)
graph.addEdge(3, a, d)
graph.addEdge(2, d, c)
graph.addEdge(8, b, e)
graph.addEdge(4, d, f)
graph.addEdge(2, c, g)
graph.addEdge(5, e, f)
graph.addEdge(5, f, g)
graph.addEdge(1, g, h)
graph.addEdge(10, e, h)

graph.print()

path = aStar(graph, a, h)
print("Path weight:", path.weight)
path.print(graph)

File: CPP/AStarVincent.cpp
#include <vector>
#include <queue>
#include <string>
#include <iostream>
#include <limits>
#include <set>
#include <algorithm>

struct Node;

struct Edge{
    int weight = 0;
    int node1;
    int node2;
    Edge(int weight, int node1, int node2) : weight(weight), node1(node1), node2(node2){}
};

struct Node{
    std::string name{};
    int index;
    double gCost = std::numeric_limits<double>::infinity();
    double hCost = 0;

    int parent = -1;

    Node(){}
    Node(std::string name, double hCost) : name(name), hCost(hCost){}

    void SetGCost(double gCost){
        this->gCost = gCost;
    }

    void SetParent(int parent){
        this->parent = parent;
    }

    void SetIndex(int index){
        this->index = index;
    }

    bool operator<(const Node& other) const{
        return gCost + hCost < other.gCost + other.hCost;
    }
};

struct CompareNode {
    bool operator()(Node* const& p1, Node* const& p2) {
        // return "true" if "p1" is ordered before "p2", for example:
        return p1->gCost + p1->hCost > p2->gCost + p2->hCost;
    }
};

struct Graph{
    std::vector<Node> nodes{};
    std::vector<Edge> edges{};

    int AddNode(std::string name, double hCost){
        nodes.emplace_back(name, hCost);
        nodes.back().SetIndex(nodes.size() - 1);
        return nodes.size() - 1;
    }

    void AddEdge(int node1, int node2, int weight){
        std::cout << "Adding edge between " << nodes[node1].name << " and " << nodes[node2].name << std::endl;

        edges.emplace_back(weight, node1, node2);
        edges.emplace_back(weight, node2, node1);
    }

    void Print(){
        for(int i = 0; i < nodes.size(); i++){
            std::cout << nodes[i].name << ": ";
            for(Edge* edge : GetEdges(i)){
                std::cout << nodes[edge->node2].name << " ";
            }
            std::cout << std::endl;
        }
    }

    std::vector<Edge*> GetEdges(int node){
        std::vector<Edge*> result{};
        for(Edge& edge : edges){
            if(edge.node1 == node){
                result.emplace_back(&edge);
            }
        }
        return result;
    }

    Node& GetNode(int node){
        return nodes[node];
    }
};

struct Path{
    std::vector<Node*> nodes;
    int weight;

    void Print(){
        for(Node* node : nodes){
            if (node != nullptr) {
                std::cout << node->name << " ";
            }
        }
        std::cout << std::endl;
    }
};

Path AStar(Graph& graph, int start, int end){
    std::priority_queue<Node*, std::vector<Node*>, CompareNode> queue;
    std::set<Node*> visited;
    Node& startNode = graph.GetNode(start);
    startNode.SetGCost(0);
    queue.push(&startNode);

    Node* current;
    
    while(!queue.empty()){
        current = queue.top();
        queue.pop();
        visited.insert(current);

        if(current->index == end){
            break;
        }

        for(Edge* edge : graph.GetEdges(current->index)){
            Node& node = graph.GetNode(edge->node2);
            if(visited.find(&node) != visited.end()){
                continue;
            }
            double gCost = current->gCost + edge->weight;
            if (gCost < node.gCost) {
                node.SetGCost(gCost);
                node.SetParent(current->index);
                queue.push(&node);
            }
        }
    }

    Path path;
    path.weight = current->gCost;
    while(current->index != start){
        path.nodes.emplace_back(current);
        current = &graph.GetNode(current->parent);
    }
    path.nodes.emplace_back(&graph.GetNode(start));

    std::reverse(path.nodes.begin(), path.nodes.end());

    return path;
}

int main(){
    Graph graph{};
    int a = graph.AddNode("A", 10);
    int b = graph.AddNode("B", 3);
    int c = graph.AddNode("C", 5);
    int d = graph.AddNode("D", 6);
    int e = graph.AddNode("E", 8);
    int f = graph.AddNode("F", 4);
    int g = graph.AddNode("G", 2);
    int h = graph.AddNode("H", 0);

    graph.AddEdge(a, b, 2);
    graph.AddEdge(a, c, 10);
    graph.AddEdge(a, d, 3);
    graph.AddEdge(d, c, 2);
    graph.AddEdge(b, e, 8);
    graph.AddEdge(d, f, 4);
    graph.AddEdge(c, g, 2);
    graph.AddEdge(f, e, 5);
    graph.AddEdge(f, g, 5);
    graph.AddEdge(g, h, 1);
    graph.AddEdge(e, h, 10);

    graph.Print();


    int start = a;
    int end = h;
    Path path = AStar(graph, start, end);

    std::cout << "Path weight: " << path.weight << std::endl;
    std::cout << "Path: ";
    path.Print();

    return 0;
}
